src/QsCompiler/Core/StatementTransformation.fs

```diff
@@ -25,63 +28,67 @@ type StatementKindTransformationBase internal (unsafe) =
+    new (options : TransformationOptions) as this = 
+        new StatementKindTransformationBase(options, "unsafe") then
+            let expressionTransformation = new ExpressionTransformationBase(options)
+            let statementTransformation = new StatementTransformationBase((fun _ -> this), (fun _ -> this.Expressions), options)
             this.ExpressionTransformationHandle <- fun _ -> expressionTransformation
             this.StatementTransformationHandle <- fun _ -> statementTransformation
 
+    new (statementTransformation : unit -> StatementTransformationBase, expressionTransformation : unit -> ExpressionTransformationBase) =
+        new StatementKindTransformationBase(statementTransformation, expressionTransformation, TransformationOptions.Default)
 
-    abstract member onQubitInitializer : ResolvedInitializer -> ResolvedInitializer 
-    default this.onQubitInitializer init = 
-        match init.Resolution with 
-        | SingleQubitAllocation      -> SingleQubitAllocation
-        | QubitRegisterAllocation ex -> QubitRegisterAllocation (this.Expressions.Transform ex)
-        | QubitTupleAllocation is    -> QubitTupleAllocation ((is |> Seq.map this.onQubitInitializer).ToImmutableArray())
-        | InvalidInitializer         -> InvalidInitializer
-        |> ResolvedInitializer.New 
+    new () = new StatementKindTransformationBase(TransformationOptions.Default)
 
+
+    // methods invoked before selective statements
+    
     abstract member beforeVariableDeclaration : SymbolTuple -> SymbolTuple
     default this.beforeVariableDeclaration syms = syms
 
+
+    // subconstructs used within statements 
+
     abstract member onSymbolTuple : SymbolTuple -> SymbolTuple
     default this.onSymbolTuple syms = syms
 
+    abstract member onQubitInitializer : ResolvedInitializer -> ResolvedInitializer 
+    default this.onQubitInitializer init = 
+        let transformed = init.Resolution |> function 
+            | SingleQubitAllocation              -> SingleQubitAllocation
+            | QubitRegisterAllocation ex as orig -> QubitRegisterAllocation |> Node.BuildOr orig (this.Expressions.Transform ex)
+            | QubitTupleAllocation is as orig    -> QubitTupleAllocation |> Node.BuildOr orig (is |> Seq.map this.onQubitInitializer |> ImmutableArray.CreateRange)
+            | InvalidInitializer                 -> InvalidInitializer
+        ResolvedInitializer.New |> Node.BuildOr init transformed
 
-    abstract member onExpressionStatement : TypedExpression -> QsStatementKind
-    default this.onExpressionStatement ex = this.Expressions.Transform ex |> QsExpressionStatement
+    abstract member onPositionedBlock : QsNullable<TypedExpression> * QsPositionedBlock -> QsNullable<TypedExpression> * QsPositionedBlock
+    default this.onPositionedBlock (intro : QsNullable<TypedExpression>, block : QsPositionedBlock) = 
+        let location = this.Statements.onLocation block.Location
```

Comment:
Shouldn't this check `options.Disable` before recursing? And similarly in other methods in this class.